File: articles/parenting-family-systems-and-human-development/python/model_family_systems.py
#!/usr/bin/env python3
"""Model synthetic parenting, family systems, and development."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "family_systems_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_family_systems_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    development_model = smf.ols(
        """
        development_score ~ lag_score + time + current_parenting +
        current_family + current_stress + household_stability +
        kin_support + economic_security + current_sibling +
        current_regulation + caregiver_support +
        current_parenting:current_family
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    family_support_model = smf.ols(
        """
        family_support_index ~ time + current_parenting + current_family +
        household_stability + kin_support + economic_security +
        current_sibling + current_regulation + caregiver_support -
        current_stress
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_family_systems_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(development_model.summary().as_text())
        f.write("\n\nFAMILY SUPPORT INDEX MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(family_support_model.summary().as_text())

    trajectory = panel.groupby(["time", "caregiver_support"], as_index=False).agg(
        average_development=("development_score", "mean"),
        average_family_support=("family_support_index", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory["group"] = trajectory["caregiver_support"].map({0: "No added support", 1: "Caregiver support"})
    trajectory.to_csv(OUTPUTS_DIR / "python_development_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("group"):
        plt.plot(group["time"], group["average_development"], marker="o", label=group_name)
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Parenting, Family Systems, and Development")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_development_trajectory.png", dpi=160)
    plt.close()

    support_trajectory = panel.groupby("time", as_index=False).agg(
        average_family_support=("family_support_index", "mean"),
        average_parenting=("current_parenting", "mean"),
        average_family=("current_family", "mean"),
        average_stress=("current_stress", "mean"),
    )
    support_trajectory.to_csv(OUTPUTS_DIR / "python_family_support_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(support_trajectory["time"], support_trajectory["average_family_support"], marker="o")
    plt.xlabel("Time")
    plt.ylabel("Average family support index")
    plt.title("Synthetic Family Support Trajectory")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_family_support_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("family_support_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_development=("development_score", "mean"),
        average_family_support=("family_support_index", "mean"),
        average_parenting=("current_parenting", "mean"),
        average_family=("current_family", "mean"),
        average_stress=("current_stress", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_family_support_profiles.csv", index=False)

    print("Wrote Python family systems model outputs.")


if __name__ == "__main__":
    main()

File: articles/parenting-family-systems-and-human-development/r/model_family_systems.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "family_systems_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_family_systems_panel.py first.")

panel <- read.csv(data_path)

development_model <- lmer(
  development_score ~ time + current_parenting + current_family +
    current_stress + household_stability + kin_support +
    economic_security + current_sibling + current_regulation +
    caregiver_support + current_parenting:current_family +
    (1 + time | household_id/child_id),
  data = panel
)

support_model <- lmer(
  family_support_index ~ time + current_parenting + current_family +
    current_stress + household_stability + kin_support +
    economic_security + current_sibling + current_regulation +
    caregiver_support +
    (1 + time | household_id/child_id),
  data = panel
)

capture.output(
  list(
    development_model = summary(development_model),
    family_support_model = summary(support_model)
  ),
  file = file.path(outputs_dir, "r_family_systems_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, caregiver_support) |>
  summarize(
    average_development = mean(development_score),
    average_family_support = mean(family_support_index),
    standard_error = sd(development_score) / sqrt(n()),
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    .groups = "drop"
  ) |>
  mutate(group = ifelse(caregiver_support == 1, "Caregiver support", "No added support"))

write.csv(trajectory, file.path(outputs_dir, "r_development_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_development_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development, linetype = group)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = group), alpha = 0.12) +
    labs(title = "Synthetic Parenting, Family Systems, and Development", x = "Time", y = "Average development", linetype = "Group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

family_support <- panel |>
  group_by(time) |>
  summarize(
    average_family_support = mean(family_support_index),
    average_parenting = mean(current_parenting),
    average_family = mean(current_family),
    average_stress = mean(current_stress),
    .groups = "drop"
  )

write.csv(family_support, file.path(outputs_dir, "r_family_support_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_family_support_trajectory.png"),
  plot = ggplot(family_support, aes(x = time, y = average_family_support)) +
    geom_line(linewidth = 1) +
    geom_point() +
    labs(title = "Synthetic Family Support Trajectory", x = "Time", y = "Average family support") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R family systems outputs.")
